modules/serverpod_auth/serverpod_auth_server/lib/src/business/sms_auth.dart

```diff
@@ -0,0 +1,156 @@
+      'The sendValidationEmail property needs to be set in AuthConfig.',
+    );
+
+    var otp = _generateOTP();
+
+    var hash = generateHashData(
+      phoneNumber: phoneNumber,
+      otp: otp,
+    );
+    await AuthConfig.current.sendValidationSMS!(
+      phoneNumber,
+      otp,
+    );
+    return hash;
+  }
+
+  /// Verifies that the OTP is correct and that the hash hasn't been tampered
+  /// with. If the OTP is correct, the user is logged in and a session is
+  /// returned.
+  static Future<bool> verifyOTPRequest({
+    required String phoneNumber,
+    required String otp,
+    required String storedHash,
+  }) async {
+    var verification = verifyHashData(
+      hashedData: storedHash,
+      phoneNumber: phoneNumber,
+      otp: otp,
+    );
+
+    if (!verification) {
+      return false;
+    }
+
+    return true;
+  }
+
+  /// Creates a new user with the given phone number.
+  static Future<SmsAuth?> createUser(
+    Session session,
+    String phoneNumber,
+  ) async {
+    var userInfo = await Users.findUserByPhoneNumber(session, phoneNumber);
+
+    if (userInfo == null) {
+      userInfo = UserInfo(
+        userIdentifier: phoneNumber,
+        phoneNumber: phoneNumber,
+        userName: phoneNumber,
+        created: DateTime.now(),
+        scopeNames: [],
+        blocked: false,
+      );
+      session.log('creating user', level: LogLevel.debug);
+      userInfo = await Users.createUser(session, userInfo, 'SMS');
+      if (userInfo == null) return null;
+    }
+
+    session.log('creating SMS auth', level: LogLevel.debug);
+    var auth = SmsAuth(
+      userId: userInfo.id!,
+      phoneNumber: phoneNumber,
+    );
+    await session.db.insert(auth);
+
+    await UserImages.setDefaultUserImage(session, userInfo.id!);
+    await Users.invalidateCacheForUser(session, userInfo.id!);
+    userInfo = await Users.findUserByUserId(session, userInfo.id!);
+    if (userInfo == null) {
+      session.log('failed to find user', level: LogLevel.error);
+      return null;
+    }
+
+    session.log('returning created user', level: LogLevel.debug);
+    return auth;
+  }
+
+  /// Generates a new hash from a users phone number, OTP and secret hash key.
+  /// This value will be used to verify the OTP when the user sends it without.
+  /// Without having the risk of storing the OTP in the database.
+  static String generateHashData(
+      {required String phoneNumber,
+      required String otp,
+      DateTime? expirationTime}) {
+    var secretHashKey =
+        Serverpod.instance.getPassword('SMSHash') ?? 'serverpod SMS hash';
+    expirationTime ??= DateTime.now().add(AuthConfig.current.smsExpirationTime);
+
+    var dataToHash = '$phoneNumber$otp$secretHashKey';
+    var hmac = Hmac(sha256, utf8.encode(secretHashKey));
+    var digest = hmac.convert(utf8.encode(dataToHash));
+    var hashedData = digest.toString();
+
+    var hash = '$hashedData@${expirationTime.toIso8601String()}';
```

Comment:
Suggestion: Add the experiation time to the hashed data.

If the expiration time is not verifiable for the server an attacker could easily change it to bypass the max expiration time.